shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java

```diff
@@ -186,29 +190,33 @@ public boolean onTouchEvent(@NonNull MotionEvent event, @NonNull Matrix transfor
   public boolean onGenericMotionEvent(@NonNull MotionEvent event) {
     // Method isFromSource is only available in API 18+ (Jelly Bean MR2)
     // Mouse hover support is not implemented for API < 18.
+
+    Log.v(TAG, "onGenericMotionEvent, event " + event);
+
     boolean isPointerEvent =
         Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2
             && event.isFromSource(InputDevice.SOURCE_CLASS_POINTER);
+
     boolean isMovementEvent =
         (event.getActionMasked() == MotionEvent.ACTION_HOVER_MOVE
             || event.getActionMasked() == MotionEvent.ACTION_SCROLL);
-    if (!isPointerEvent || !isMovementEvent) {
-      return false;
-    }
-
-    int pointerChange = getPointerChangeForAction(event.getActionMasked());
-    ByteBuffer packet =
-        ByteBuffer.allocateDirect(
-            event.getPointerCount() * POINTER_DATA_FIELD_COUNT * BYTES_PER_FIELD);
-    packet.order(ByteOrder.LITTLE_ENDIAN);
 
-    // ACTION_HOVER_MOVE always applies to a single pointer only.
-    addPointerForIndex(event, event.getActionIndex(), pointerChange, 0, IDENTITY_TRANSFORM, packet);
-    if (packet.position() % (POINTER_DATA_FIELD_COUNT * BYTES_PER_FIELD) != 0) {
-      throw new AssertionError("Packet position is not on field boundary.");
+    if (isPointerEvent || isMovementEvent) {
```

Comment:
make touchProcessor.onGenericMotionEvent in unit test work.

MotionEvent.isFromSource() is not mockable because it's from super class in InputEvent.For the code before tha patch:

   boolean isPointerEvent =
       Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2
           && event.isFromSource(InputDevice.SOURCE_CLASS_POINTER);                   ======> Here always false.

   if (!isPointerEvent || !isMovementEvent) {    ========================> always true here 
     return false;
    }

  renderer.dispatchPointerDataPacket(packet, packet.position()); =========> always skipped. so that test case always end up with "wanted but not invoked" erro...

For code after the patch:

if (isPointerEvent || isMovementEvent) { 

so event if isPointerEvent return false, isMovementEvent return true. Test case works.

Reply:
The code here is supposed to return false, when we don't handle the event. Now you've changed it, so it will now try to handle events with `!isPointerEvent`. Can you clarify - is this change needed for your special device input events, or just for testing? Because I hope we can find some other way to fix the test, it may break or cause weird behaviour in the field to start handling these generic events.

Reply:
Can you make it more clear here, use the same format as before, so the difference is more obvious. So it would be `!isPointerEvent && !isMovementEvent` to `return false` at the beginning. Then no need to indent the rest of the function.

Reply:
You can just simply revert this patch

-    if (!isPointerEvent || !isMovementEvent) {
-      return false;
-    }

and do the test with "testing/run_tests.py --type=java"

The unexpectedPointerChange test case will always failed with the error below :

org.mockito.exceptions.verification.VerificationInOrderFailure: 
Verification in order failure
Wanted but not invoked:
mockRenderer.dispatchPointerDataPacket(
    <Capturing argument>,
    <Capturing argument>
);
-> at io.flutter.embedding.engine.renderer.FlutterRenderer.dispatchPointerDataPacket(FlutterRenderer.java:484)
Wanted anywhere AFTER following interaction:
mockRenderer.dispatchPointerDataPacket(
    java.nio.DirectByteBuffer[pos=280 lim=280 cap=280],
    280
);

just because event.isFromSource will alway return false. Event though I have set 

when(event.isFromSource()).thenReturn(true)  in mockEvent , It doest not work either. 

I think the reason is isFromSource() is not mockable because it's from super class in InputEvent.

So I have to chage this code.